plugins/transport-nio/src/main/java/org/elasticsearch/http/nio/cors/NioCorsHandler.java

```diff
@@ -0,0 +1,235 @@
+        if (setOrigin(response)) {
+            setAllowMethods(response);
+            setAllowHeaders(response);
+            setAllowCredentials(response);
+            setMaxAge(response);
+            setPreflightHeaders(response);
+            ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
+        } else {
+            forbidden(ctx, request);
+        }
+    }
+
+    private static void forbidden(final ChannelHandlerContext ctx, final HttpRequest request) {
+        ctx.writeAndFlush(new DefaultFullHttpResponse(request.protocolVersion(), HttpResponseStatus.FORBIDDEN))
+            .addListener(ChannelFutureListener.CLOSE);
+    }
+
+    private static boolean isSameOrigin(final String origin, final String host) {
+        if (Strings.isNullOrEmpty(host) == false) {
+            // strip protocol from origin
+            final String originDomain = SCHEME_PATTERN.matcher(origin).replaceFirst("");
+            if (host.equals(originDomain)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * This is a non CORS specification feature which enables the setting of preflight
+     * response headers that might be required by intermediaries.
+     *
+     * @param response the HttpResponse to which the preflight response headers should be added.
+     */
+    private void setPreflightHeaders(final HttpResponse response) {
+        response.headers().add(config.preflightResponseHeaders());
+    }
+
+    private boolean setOrigin(final HttpResponse response) {
+        final String origin = request.headers().get(HttpHeaderNames.ORIGIN);
+        if (!Strings.isNullOrEmpty(origin)) {
+            if ("null".equals(origin) && config.isNullOriginAllowed()) {
+                setAnyOrigin(response);
+                return true;
+            }
+
+            if (config.isAnyOriginSupported()) {
+                if (config.isCredentialsAllowed()) {
+                    echoRequestOrigin(response);
+                    setVaryHeader(response);
+                } else {
+                    setAnyOrigin(response);
+                }
+                return true;
+            }
+            if (config.isOriginAllowed(origin)) {
+                setOrigin(response, origin);
+                setVaryHeader(response);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean validateOrigin() {
+        if (config.isAnyOriginSupported()) {
+            return true;
+        }
+
+        final String origin = request.headers().get(HttpHeaderNames.ORIGIN);
+        if (Strings.isNullOrEmpty(origin)) {
+            // Not a CORS request so we cannot validate it. It may be a non CORS request.
+            return true;
+        }
+
+        if ("null".equals(origin) && config.isNullOriginAllowed()) {
+            return true;
```

Comment:
I think this should be something more like:
```
if ("null".equals(origin)) {
  return config.isNullOriginAllowed();
}
```
I don't think it makes sense to fall through to pattern checking the origin is null, and null is not allowed.
That is, if we have an explicit config option for `null` handling, then that should be the only way to determine whether null is allowed.

Reply:
I actually do not think this is correct based on how we have implemented Cors. Our setup, means that `config.isNullOriginAllowed` will always be false. We never modify that config. We only allow configuring `config.isOriginAllowed(...)` through `http.cors.allow-origin`. So if the user wants to allow null, that must be in the regex. We will only evaluate that regex if we fall through.

Reply:
I have a (non-blocking) concern about copying code across with the obligations to maintain it, but only use part of it.
Is there an argument for why we're keeping parts of this that we don't actually use?

Reply:
I'm not sure why we took the approach we did. It is on my roadmap to change at some point.

I have started playing around with implementing Cors on our side opposed to Netty's side (https://github.com/tbrooks8/elasticsearch/tree/cors_without_netty). Essentially we would do the Cors logic AFTER we have turned a netty http request into a `RestRequest`. And then nio / netty4 modules would use the same code base (as the logic would live in `server`). If that turns out to be too complicated, I can explore paring down what we use from netty.